File: app-mobile/src/main/java/org/jrs82/fsclock/mobile/MobileWidgetOrderActivity.java
package org.jrs82.fsclock.mobile;

import android.content.SharedPreferences;
import android.graphics.Color;
import android.os.Bundle;
import android.view.Gravity;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;
import androidx.preference.PreferenceManager;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import org.jrs82.fsclock.R;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Etusivun widgettien järjestely raahaamalla (drag & drop) sekä näkyvyyden
 * valinta checkboxilla. Korvaa aiemman ylös/alas-nappidialogin. Lista sisältää
 * kiinteät widgetit, yhdistetyn uutiswidgetin ja jokaisen uutislähteen
 * per-lähde-widgettinä.
 */
public class MobileWidgetOrderActivity extends AppCompatActivity {

    private SharedPreferences prefs;
    private final List<Row> rows = new ArrayList<>();
    private Adapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        MobileThemeController.apply(this);
        super.onCreate(savedInstanceState);
        setTitle("Etusivun kortit");
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }
        prefs = PreferenceManager.getDefaultSharedPreferences(this);

        RecyclerView recycler = new RecyclerView(this);
        recycler.setLayoutManager(new LinearLayoutManager(this));
        recycler.setBackgroundColor(themeColor(android.R.attr.colorBackground, Color.BLACK));
        int pad = dp(8);
        recycler.setPadding(pad, pad, pad, pad);
        recycler.setClipToPadding(false);
        setContentView(recycler);

        buildRows();
        adapter = new Adapter();
        recycler.setAdapter(adapter);

        touchHelper = new ItemTouchHelper(new DragCallback());
        touchHelper.attachToRecyclerView(recycler);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void buildRows() {
        rows.clear();
        for (String id : loadOrder()) {
            rows.add(new Row(id, widgetTitle(id), isVisible(id), hasToggle(id)));
        }
    }

    /** Tallennettu järjestys + kaikki tunnetut etusivun kortit jotka puuttuvat. */
    private List<String> loadOrder() {
        List<String> order = new ArrayList<>();
        String raw = prefs.getString(ComposeHomeWidgetsKt.KEY_HOME_ORDER, null);
        if (raw != null) {
            for (String token : raw.split(",")) {
                String id = token.trim();
                if (isKnown(id) && !order.contains(id)) order.add(id);
            }
        }
        for (String id : allKnownWidgets()) {
            if (!order.contains(id)) order.add(id);
        }
        return order;
    }

    /** Kaikki tunnetut etusivun kortit: kiinteät + per-lähde-uutiskortit (Kotlin-malli). */
    private List<String> allKnownWidgets() {
        return ComposeHomeWidgetsKt.allHomeWidgetIds(prefs);
    }

    private boolean isKnown(String id) {
        return id != null && allKnownWidgets().contains(id);
    }

    private void persist() {
        StringBuilder order = new StringBuilder();
        SharedPreferences.Editor editor = prefs.edit();
        for (Row r : rows) {
            if (order.length() > 0) order.append(',');
            order.append(r.id);
            String key = visibilityKey(r.id);
            if (key != null) editor.putBoolean(key, r.visible);
        }
        editor.putString(ComposeHomeWidgetsKt.KEY_HOME_ORDER, order.toString());
        editor.apply();
    }

    // ---- Näkyvyys ----

    /** Kaikki etusivun kortit voi piilottaa. */
    private boolean hasToggle(String id) {
        return true;
    }

    private String visibilityKey(String id) {
        return ComposeHomeWidgetsKt.KEY_HOME_SHOW_PREFIX + id;
    }

    private boolean defaultVisible(String id) {
        return ComposeHomeWidgetsKt.defaultVisibleForId(id);
    }

    private boolean isVisible(String id) {
        return prefs.getBoolean(visibilityKey(id), defaultVisible(id));
    }

    private String widgetTitle(String id) {
        return ComposeHomeWidgetsKt.homeWidgetTitleForId(prefs, id);
    }

    private int themeColor(int attr, int fallback) {
        android.util.TypedValue tv = new android.util.TypedValue();
        if (getTheme().resolveAttribute(attr, tv, true)) return tv.data;
        return fallback;
    }

    private int dp(int value) {
        return Math.round(value * getResources().getDisplayMetrics().density);
    }

    // ---- Data ----

    private static final class Row {
        final String id;
        final String title;
        boolean visible;
        final boolean hasToggle;

        Row(String id, String title, boolean visible, boolean hasToggle) {
            this.id = id;
            this.title = title;
            this.visible = visible;
            this.hasToggle = hasToggle;
        }
    }

    // ---- Adapter ----

    private final class Adapter extends RecyclerView.Adapter<VH> {
        @NonNull
        @Override
        public VH onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            LinearLayout row = new LinearLayout(MobileWidgetOrderActivity.this);
            row.setOrientation(LinearLayout.HORIZONTAL);
            row.setGravity(Gravity.CENTER_VERTICAL);
            row.setBackgroundResource(R.drawable.mobile_card_bg);
            LinearLayout.LayoutParams lp = new LinearLayout.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
            lp.topMargin = dp(6);
            lp.bottomMargin = dp(6);
            row.setLayoutParams(lp);

            ImageView handle = new ImageView(MobileWidgetOrderActivity.this);
            handle.setImageResource(R.drawable.mobile_ic_drag_handle_24);
            handle.setColorFilter(ContextCompat.getColor(
                    MobileWidgetOrderActivity.this, R.color.mobile_text_muted));
            LinearLayout.LayoutParams hlp = new LinearLayout.LayoutParams(dp(28), dp(28));
            hlp.setMarginEnd(dp(12));
            handle.setLayoutParams(hlp);
            row.addView(handle);

            TextView title = new TextView(MobileWidgetOrderActivity.this);
            title.setTextColor(ContextCompat.getColor(
                    MobileWidgetOrderActivity.this, R.color.mobile_text_primary));
            title.setTextSize(16f);
            title.setLayoutParams(new LinearLayout.LayoutParams(
                    0, ViewGroup.LayoutParams.WRAP_CONTENT, 1f));
            row.addView(title);

            CheckBox check = new CheckBox(MobileWidgetOrderActivity.this);
            row.addView(check);

            return new VH(row, handle, title, check);
        }

        @Override
        public void onBindViewHolder(@NonNull VH holder, int position) {
            Row row = rows.get(position);
            holder.title.setText(row.title);
            holder.check.setOnCheckedChangeListener(null);
            holder.check.setChecked(row.visible);
            holder.check.setEnabled(row.hasToggle);
            holder.check.setVisibility(row.hasToggle ? View.VISIBLE : View.INVISIBLE);
            holder.check.setOnCheckedChangeListener((buttonView, isChecked) -> {
                int pos = holder.getBindingAdapterPosition();
                if (pos == RecyclerView.NO_POSITION) return;
                rows.get(pos).visible = isChecked;
                persist();
            });
            holder.handle.setOnTouchListener((v, event) -> {
                if (event.getActionMasked() == android.view.MotionEvent.ACTION_DOWN) {
                    dragStarter.startDrag(holder);
                }
                return false;
            });
        }

        @Override
        public int getItemCount() {
            return rows.size();
        }
    }

    private static final class VH extends RecyclerView.ViewHolder {
        final ImageView handle;
        final TextView title;
        final CheckBox check;

        VH(@NonNull View itemView, ImageView handle, TextView title, CheckBox check) {
            super(itemView);
            this.handle = handle;
            this.title = title;
            this.check = check;
        }
    }

    // ItemTouchHelper-callback drag-handlen käynnistämiseen.
    private interface DragStarter { void startDrag(RecyclerView.ViewHolder vh); }
    private DragStarter dragStarter = vh -> { };

    private final class DragCallback extends ItemTouchHelper.Callback {
        DragCallback() {
            dragStarter = MobileWidgetOrderActivity.this::beginDrag;
        }

        @Override
        public boolean isLongPressDragEnabled() {
            return false; // raahaus vain kahvasta
        }

        @Override
        public int getMovementFlags(@NonNull RecyclerView recyclerView,
                                    @NonNull RecyclerView.ViewHolder viewHolder) {
            return makeMovementFlags(
                    ItemTouchHelper.UP | ItemTouchHelper.DOWN, 0);
        }

        @Override
        public boolean onMove(@NonNull RecyclerView recyclerView,
                              @NonNull RecyclerView.ViewHolder viewHolder,
                              @NonNull RecyclerView.ViewHolder target) {
            int from = viewHolder.getBindingAdapterPosition();
            int to = target.getBindingAdapterPosition();
            if (from == RecyclerView.NO_POSITION || to == RecyclerView.NO_POSITION) return false;
            Collections.swap(rows, from, to);
            adapter.notifyItemMoved(from, to);
            return true;
        }

        @Override
        public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int direction) { }

        @Override
        public void clearView(@NonNull RecyclerView recyclerView,
                              @NonNull RecyclerView.ViewHolder viewHolder) {
            super.clearView(recyclerView, viewHolder);
            persist();
        }
    }

    private ItemTouchHelper touchHelper;

    private void beginDrag(RecyclerView.ViewHolder vh) {
        if (touchHelper != null) touchHelper.startDrag(vh);
    }
}

File: app-mobile/src/main/java/org/jrs82/fsclock/mobile/MobileWidgetOrderActivity.kt
package org.jrs82.fsclock.mobile

import android.content.SharedPreferences
import android.os.Bundle
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.compose.foundation.gestures.detectDragGestures
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.Checkbox
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.SnackbarDuration
import androidx.compose.material3.SnackbarHost
import androidx.compose.material3.SnackbarHostState
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.key
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.rememberUpdatedState
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.hapticfeedback.HapticFeedbackType
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.unit.dp
import androidx.compose.ui.zIndex
import androidx.preference.PreferenceManager
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import org.jrs82.fsclock.R
import java.util.Collections

/**
 * Etusivun korttien järjestely raahaamalla (drag & drop kahvasta) ja näkyvyyden valinta
 * checkboxilla — Compose-versio, korvaa aiemman View/RecyclerView+ItemTouchHelper-toteutuksen.
 * Luokkanimi, manifest-merkintä ja asetusten Intent-kytkentä ennallaan. Lukee ja tallentaa
 * SAMAT SharedPreferences-avaimet ([KEY_HOME_ORDER], [KEY_HOME_SHOW_PREFIX]) samalla
 * logiikalla ([allHomeWidgetIds], [homeWidgetTitleForId], [defaultVisibleForId]).
 *
 * Palaute: raahaus alkaa/pudottaa → haptinen napsahdus; pudotuksen JÄLKEEN lyhyt snackbar
 * "Järjestys tallennettu" (näkyvyyden vaihto: "Tallennettu"). Snackbar korvautuu — ei jonoa.
 * Tallennus pysyy automaattisena (ei erillistä Tallenna-painiketta).
 */
class MobileWidgetOrderActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        MobileThemeController.apply(this)
        enableEdgeToEdge()
        super.onCreate(savedInstanceState)
        setContent {
            ArkikeskusTheme(dynamicColor = MobileThemeController.dynamicColor(this)) {
                WidgetOrderScreen(onBack = { finish() })
            }
        }
    }
}

private const val ROW_HEIGHT_DP = 64

private class WidgetRowData(val id: String, val title: String, visible: Boolean) {
    var visible by mutableStateOf(visible)
}

/** Kortin oma ikoni (sama jota kortti käyttää etusivulla) — rivin tunnistettavuus. */
private fun homeWidgetIconForId(id: String): Int = when {
    id.startsWith("news:") -> R.drawable.mobile_ic_news_24 // per-lähde-uutiskortti
    else -> when (id) {
        "clock" -> R.drawable.mobile_ic_clock_24
        "holiday" -> R.drawable.mobile_ic_flag_24
        "weather" -> R.drawable.mobile_ic_weather_24
        "electricity" -> R.drawable.mobile_ic_bolt_24
        "warnings" -> R.drawable.mobile_ic_warning_24
        "sensors" -> R.drawable.mobile_ic_thermometer_24
        "traffic" -> R.drawable.mobile_ic_transit_tram
        "news", "news_foreign" -> R.drawable.mobile_ic_news_24
        "transit" -> R.drawable.mobile_ic_bus_24
        else -> R.drawable.mobile_ic_dashboard_24
    }
}

/** Tallennettu järjestys + kaikki tunnetut kortit jotka puuttuvat (sama kuin View-versio). */
private fun loadWidgetOrder(prefs: SharedPreferences): List<String> {
    val known = allHomeWidgetIds(prefs)
    val order = ArrayList<String>()
    prefs.getString(KEY_HOME_ORDER, null)?.split(",")?.forEach { token ->
        val id = token.trim()
        if (known.contains(id) && !order.contains(id)) order.add(id)
    }
    known.forEach { if (!order.contains(it)) order.add(it) }
    return order
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun WidgetOrderScreen(onBack: () -> Unit) {
    val context = LocalContext.current
    val prefs = remember { PreferenceManager.getDefaultSharedPreferences(context) }
    val rows = remember {
        mutableStateListOf<WidgetRowData>().apply {
            loadWidgetOrder(prefs).forEach { id ->
                add(
                    WidgetRowData(
                        id,
                        homeWidgetTitleForId(prefs, id),
                        prefs.getBoolean(KEY_HOME_SHOW_PREFIX + id, defaultVisibleForId(id)),
                    ),
                )
            }
        }
    }

    val haptic = LocalHapticFeedback.current
    val scope = rememberCoroutineScope()
    val snackbarHostState = remember { SnackbarHostState() }
    var snackbarJob by remember { mutableStateOf<Job?>(null) }

    fun persist() {
        val editor = prefs.edit()
        editor.putString(KEY_HOME_ORDER, rows.joinToString(",") { it.id })
        rows.forEach { editor.putBoolean(KEY_HOME_SHOW_PREFIX + it.id, it.visible) }
        editor.apply()
    }

    // Lyhyt (~1,5 s) snackbar joka KORVAA edellisen (ei jonoa): peruu aiemman työn ja näyttää uuden.
    fun showSaved(message: String) {
        snackbarJob?.cancel()
        snackbarJob = scope.launch {
            val shown = launch {
                snackbarHostState.showSnackbar(message, duration = SnackbarDuration.Indefinite)
            }
            delay(1500)
            shown.cancel()
        }
    }

    val rowHeightPx = with(LocalDensity.current) { ROW_HEIGHT_DP.dp.toPx() }
    var draggingIndex by remember { mutableIntStateOf(-1) }
    var dragOffset by remember { mutableFloatStateOf(0f) }
    var didReorder by remember { mutableStateOf(false) }

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Etusivun kortit") },
                navigationIcon = {
                    IconButton(onClick = onBack) {
                        Icon(painterResource(R.drawable.mobile_ic_arrow_back), contentDescription = "Takaisin")
                    }
                },
            )
        },
        snackbarHost = { SnackbarHost(snackbarHostState) },
    ) { padding ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(padding)
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 8.dp),
        ) {
            Text(
                "Vedä kahvasta järjestääksesi. Valinta näyttää kortin etusivulla.",
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.padding(start = 8.dp, end = 8.dp, top = 12.dp, bottom = 6.dp),
            )
            rows.forEachIndexed { index, row ->
                key(row.id) {
                    WidgetOrderRow(
                        row = row,
                        dragging = index == draggingIndex,
                        offsetY = if (index == draggingIndex) dragOffset else 0f,
                        onDragStart = {
                            draggingIndex = rows.indexOf(row)
                            dragOffset = 0f
                            didReorder = false
                            haptic.performHapticFeedback(HapticFeedbackType.TextHandleMove)
                        },
                        onDragBy = { dy ->
                            dragOffset += dy
                            // Raahaus: vaihda viereiseen kun ylitetään reilu puolikas rivistä;
                            // dragOffset korjataan rivin verran → kortti pysyy sormen alla.
                            while (dragOffset > rowHeightPx * 0.6f && draggingIndex < rows.lastIndex) {
                                Collections.swap(rows, draggingIndex, draggingIndex + 1)
                                draggingIndex++
                                dragOffset -= rowHeightPx
                                didReorder = true
                            }
                            while (dragOffset < -rowHeightPx * 0.6f && draggingIndex > 0) {
                                Collections.swap(rows, draggingIndex, draggingIndex - 1)
                                draggingIndex--
                                dragOffset += rowHeightPx
                                didReorder = true
                            }
                        },
                        onDragEnd = {
                            val reordered = didReorder
                            draggingIndex = -1
                            dragOffset = 0f
                            didReorder = false
                            // Snackbar + vahvempi napsahdus VAIN kun järjestys oikeasti muuttui
                            // (pelkkä kahvan kosketus ilman siirtoa ei tallenna turhaan).
                            if (reordered) {
                                persist()
                                haptic.performHapticFeedback(HapticFeedbackType.LongPress)
                                showSaved("Järjestys tallennettu")
                            }
                        },
                        onToggle = { v ->
                            row.visible = v
                            persist()
                            haptic.performHapticFeedback(HapticFeedbackType.TextHandleMove)
                            showSaved("Tallennettu")
                        },
                    )
                }
            }
            Spacer(Modifier.height(16.dp))
        }
    }
}

@Composable
private fun WidgetOrderRow(
    row: WidgetRowData,
    dragging: Boolean,
    offsetY: Float,
    onDragStart: () -> Unit,
    onDragBy: (Float) -> Unit,
    onDragEnd: () -> Unit,
    onToggle: (Boolean) -> Unit,
) {
    // rememberUpdatedState → pointerInput(Unit)-ele käyttää aina tuoreimpia callbackeja
    // (rivit vaihtavat paikkaa kesken eleen).
    val currentOnDragStart by rememberUpdatedState(onDragStart)
    val currentOnDragBy by rememberUpdatedState(onDragBy)
    val currentOnDragEnd by rememberUpdatedState(onDragEnd)

    // Nostettu kortti = vaaleampi pinta; piilotettu = haaleampi pinta; muuten normaali korttiväri.
    val containerColor = when {
        dragging -> MaterialTheme.colorScheme.surfaceBright
        row.visible -> MaterialTheme.colorScheme.surfaceVariant
        else -> MaterialTheme.colorScheme.surfaceContainerLow
    }
    // Piilotettu rivi himmenee (paitsi raahattaessa, jolloin se on selvästi nostettu).
    val contentAlpha = if (row.visible || dragging) 1f else 0.55f

    Card(
        modifier = Modifier
            .fillMaxWidth()
            .padding(vertical = 4.dp)
            .height(ROW_HEIGHT_DP.dp)
            .zIndex(if (dragging) 1f else 0f)
            .graphicsLayer {
                translationY = offsetY
                if (dragging) {
                    scaleX = 1.03f
                    scaleY = 1.03f
                }
            },
        colors = CardDefaults.cardColors(containerColor = containerColor),
        elevation = CardDefaults.cardElevation(defaultElevation = if (dragging) 8.dp else 1.dp),
    ) {
        Row(
            modifier = Modifier
                .fillMaxSize()
                .graphicsLayer { alpha = contentAlpha }
                .padding(horizontal = 8.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            // Kahva = ainoa raahauksen tartuntakohta (lista vierittyy normaalisti muualta).
            Box(
                modifier = Modifier
                    .size(40.dp)
                    .pointerInput(Unit) {
                        detectDragGestures(
                            onDragStart = { currentOnDragStart() },
                            onDrag = { change, amount ->
                                change.consume()
                                currentOnDragBy(amount.y)
                            },
                            onDragEnd = { currentOnDragEnd() },
                            onDragCancel = { currentOnDragEnd() },
                        )
                    },
                contentAlignment = Alignment.Center,
            ) {
                Icon(
                    painterResource(R.drawable.mobile_ic_drag_indicator_24),
                    contentDescription = "Raahaa järjestääksesi",
                    tint = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
            // Kortin oma ikoni.
            Icon(
                painterResource(homeWidgetIconForId(row.id)),
                contentDescription = null,
                tint = MaterialTheme.colorScheme.primary,
                modifier = Modifier.size(24.dp),
            )
            Spacer(Modifier.width(14.dp))
            Text(
                row.title,
                modifier = Modifier.weight(1f),
                style = MaterialTheme.typography.bodyLarge,
            )
            Checkbox(checked = row.visible, onCheckedChange = onToggle)
        }
    }
}
